### PR TITLE
Game 20 setup socket for chat

### PR DESCRIPTION
**Pull request by john-amiscaray**

### CODE REVIEW

File: Server/src/main/java/me/dulce/gamesite/gamesite2/rooms/RoomManager.java [hunk 1]
```diff
@@ -94,8 +95,8 @@ public boolean isUserInRoom(User user, Room room) {
         return room.getAllJoinedUsers().contains(user) || room.getAllSpectatingUsers().contains(user);
     }
 
-    public boolean handleIncomingRoomData(User user, GameData data) {
-        return activeRooms.get(data.roomId()).handleGameDataReceived(user, data);
+    public boolean handleIncomingRoomData(User user, GameData data, SocketController controller) {
```
Comment (KyleDulce) [inline Server/src/main/java/me/dulce/gamesite/gamesite2/rooms/RoomManager.java:98]: Couldn't you @autowire the socket controller? RoomManager is a Service here

File: Server/src/main/java/me/dulce/gamesite/gamesite2/rooms/managers/Room.java [hunk 2]
```diff
@@ -65,6 +67,33 @@ public void userLeave(User user) {
...
 
+    /**
+     * This method is for common logic among games for handling GameData such as broadcasting chat messages.
+     * A separate protected method, processGameDataForGame is called to handle game specific GameData processing.
+     * @param user The user sending the gameData
+     * @param response The GameData from the user
+     * @param controller The SocketController for broadcasting messages where needed
+     * @return A boolean indicating if the response was successful
+     */
+    public final boolean handleGameDataReceived(User user, GameData response, SocketController controller){
```
Comment (KyleDulce) [inline Server/src/main/java/me/dulce/gamesite/gamesite2/rooms/managers/Room.java:78]: Socket Controller is a singleton according to its declaration. Would it make more sense to provide it in the constructor?

**Review by KyleDulce — CHANGES_REQUESTED** [review]
Can you add unit tests (Possibly in the Room class) on testing handleGameDataReceived on if a chat message is passed through? You can possible just Mock the SocketController and verify the broadcast message method was called and the regular process is also called

File: Server/src/main/java/me/dulce/gamesite/gamesite2/rooms/RoomManager.java (hunk 1 above)
Comment (john-amiscaray) [reply Server/src/main/java/me/dulce/gamesite/gamesite2/rooms/RoomManager.java:98]: Think that would cause a circular dependency problem

**Comment by john-amiscaray** [comment]
> Can you add unit tests (Possibly in the Room class) on testing handleGameDataReceived on if a chat message is passed through? You can possible just Mock the SocketController and verify the broadcast message method was called and the regular process is also called

Whoops yeah, I did forget to test. I'll get to that.

File: Server/src/main/java/me/dulce/gamesite/gamesite2/rooms/RoomManager.java (hunk 1 above)
Comment (KyleDulce) [reply Server/src/main/java/me/dulce/gamesite/gamesite2/rooms/RoomManager.java:98]: Try lazy injection?
https://www.baeldung.com/circular-dependencies-in-spring section 4.2

File: Server/src/main/java/me/dulce/gamesite/gamesite2/rooms/managers/Room.java (hunk 2 above)
Comment (john-amiscaray) [reply Server/src/main/java/me/dulce/gamesite/gamesite2/rooms/managers/Room.java:78]: I guess you're right, I could change that

**Review by Kevinehh — APPROVED** [review]

**Comment by john-amiscaray** [comment]
Pushed a new commit for some refactoring and to address some of your comments. I'll be working on writing the tests now.

**Comment by john-amiscaray** [comment]
Committed a test for the chat message being broadcasted. I just want to also add tests for the new `SocketMessengerService` class I added. Will do that next commit.

**Review by KyleDulce — APPROVED** [review]
LGTM

**Pull request merged**